### DISCUSSION TITLE
Fixed Width Column

### DISCUSSION THREAD

**Discussion by funkpotamus-prime** [komarovalexander/ka-table#132] [Q&A]
Greetings,

First, let me say thanks for such a great component. ka-table has excellent performance and I have enjoyed the api so far. 

I recently encountered a use case that requires a fixed width column but cannot figure out how to achieve that. I have tried setting the maxWidth on the style of the column but the grid still seems to auto-size / scale the column based on available real estate.

Is is possible to fix the column size?

Cheers,

Prime

**Comment by komarovalexander** [comment]
Hi @funkpotamus-prime  
`columns.style.width` property should be a solution: https://stackblitz.com/edit/table-overview-ts-nbjntp?file=Demo.tsx

**Reply by funkpotamus-prime** [reply]
I am definitely setting the style.width property - and this works most of the time. One difference is that the table that isn't keeping the size fixed is a child table shown when expanding the parent row. Let me see if I can get a minimal example to replicate the behavior.

**Reply by funkpotamus-prime** [reply]
I think I figured it out - and it kind of makes sense. If you set the width on ALL columns, the grid appears to revert to auto-sizing. If you leave at least one column with the width unspecified, then the other columns stay fixed and the column with no width specified fills the remainder of the space. 

Demo here: https://stackblitz.com/edit/ka-table-column-width-demo?file=Demo.tsx

Leaving the last column width unspecified fixed the issue in my case.

**Reply by funkpotamus-prime** [reply]
I have a follow up question. When using the technique above (leaving the last column without a width specified), I also set the whitespace style property to "nowrap" to prevent the column from wrapping long text in a narrow column. When configured this way, the column does expand to allow enough space to show the text, but the styling on the column header does not update correctly. 
![image](https://user-images.githubusercontent.com/78034124/108638352-e011ac80-745c-11eb-811f-fd3a86255b76.png)

I have updated the prior stackblitz demo with a minimal example demonstrating the issue: https://stackblitz.com/edit/ka-table-column-width-demo?file=Demo.tsx

Is there a different way to achieve the desired effect or is this an issue that needs to be fixed with the table?

**Reply by komarovalexander** [reply]
this is how HTML table behaves because all columns have a width that is bigger than the table itself, in case if you set the table width it will be better https://stackblitz.com/edit/ka-table-column-width-demo-riucn3?file=style.scss 
but setting the column width would be a better solution here

**Reply by funkpotamus-prime** [reply]
I agree regarding setting the width. But when I do, then all of the columns have a width set and then the columns revert to auto-sizing. The behavior I would prefer to have is:

Configuring all columns with "auto-size to widest content, limited by max-width" and having a horizontal scroll bar when the total column width exceeds the size of the grid would also work. I don't really want to size the columns manually, so this is preferable.

Alternatively, if setting the width manually is needed, then the preferred behavior would be:
1. Make all columns the configured size and provide a horizontal scroll bar when the total width is wider than the table.
2. Make the last column fill the remaining space when the total width is less than the table.

Cheers,

Prime

**Reply by komarovalexander** [reply]
Hi @funkpotamus-prime 
please try this workaround:

```
  childComponents={{
    table: {
        elementAttributes: () => ({
          ref: (ref: any) => {            
            const lastThElement = ref?.querySelector('.ka-thead-cell:last-child');
            if (lastThElement 
            && !lastThElement.style.width){
              const tableWidth = ref?.offsetWidth;
              let widthDifference = tableWidth - tableProps.columns.reduce((acc, c) =>     c.style.width ? acc += c.style.width : acc, 0);
              if(widthDifference < 150){
                widthDifference = 150;                  
              }                 
              lastThElement.style.width = `${widthDifference}px`;
            }
          },
        }),
      }
  }}
```
examples
> 1. Make all columns the configured size and provide a horizontal scroll bar when the total width is wider than the table.

https://stackblitz.com/edit/ka-table-column-width-demo-2qrbux?file=Demo.tsx

> 2. Make the last column fill the remaining space when the total width is less than the table.

https://stackblitz.com/edit/ka-table-column-width-demo-paac7k?file=Demo.tsx

As for making this as autosize option: this is a great proposal, but I think I will add it later, I will wait for further requests

thanks!

**Reply by funkpotamus-prime** [reply]
That workaround successfully yields the desired behavior. Thank you for the quick response!
